core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java

```diff
@@ -58,17 +53,22 @@ private static void miscMetadataFiles(String metadataFileLocation, Set<String> m
     if (metadataFileLocation == null) {
       return;
     }
-    try {
-      TableMetadata metadata = TableMetadataParser.read(io, metadataFileLocation);
-      List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
-      List<String> previousMetadataFiles =
-          metadataLogEntries.stream().map(TableMetadata.MetadataLogEntry::file).collect(Collectors.toList());
+    TableMetadata metadata = TableMetadataParser.read(io, metadataFileLocation);
+    List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
+    List<String> previousMetadataFiles =
+        metadataLogEntries.stream().map(TableMetadata.MetadataLogEntry::file)
+            .collect(Collectors.toList());
+    if (previousMetadataFiles.size() > 0) {
       metaFiles.addAll(previousMetadataFiles);
-      if (isRecursive && previousMetadataFiles.size() > 0) {
-        miscMetadataFiles(previousMetadataFiles.get(0), metaFiles, io, isRecursive);
+      if (isRecursive) {
```

Comment:
In the above case, we will first pick the latest version(ie file 3), add all the files in file 3.
Pick the oldest exist file from the list of files in file 3, and recurse with that file.
```
      if (isRecursive) {
        for (String fileName : previousMetadataFiles) {
          // Find the first existent metadata json file and recurse
          if (io.newInputFile(fileName).exists()) {
            miscMetadataFiles(fileName, metaFiles, io, isRecursive);
            break;
          }
        }
      }
```
We dont queue up all the files present in a particular file. The max number of files queued up will be 1

Reply:
Maybe we should have that be a "findFirst" then instead of the break? To make things a bit more straightforward?

Reply:
Make sense to get the first item more explicitly. Another concern, is there a chance that order is broken?

Reply:
oh, wait, to get the first isn't correct here. Image you have v1 v2 v3 and v4, METADATA_PREVIOUS_VERSIONS_MAX is 2. You can delete v2 in v4. v2 is the first one, if you only search for the first one, you are going to miss things in v3.metadata.json, which points to v1 and v2.

Reply:
We need to find the first existent file, like in my previous logic. 
But there will be issues with that as well. There can be unreachable files like in this test case. https://github.com/apache/iceberg/pull/2564/files#diff-efb7c769faae3f19c11382d2dc4c53a69074265a87f7291f8430ad91410d48e2R141

Reply:
What is the use case of getting all metadata.json files? How about listing all *metadata.json files in the dir?

Reply:
I think we can try to load previous metadata files but give up if we cannot do that. It should be okay to leave those files too. In case if Iceberg owns the location, we will drop the entire location after that too. If not, we can leave them be.

Listing may not be an option as the same location may contain files for multiple tables. In HiveCatalog, we have table UUID as part of the file name while we don't have that in HadoopCatalog for example. 

If it is too complicated, we can drop the recursion too and I say we only clean up files currently referenced.

Reply:
Resolving this.